clean/post_officer_history.py

```diff
@@ -0,0 +1,272 @@
+                .str.replace(r"\bLa\b", "Louisiana")
+                .str.replace(r"^PlaqueminesSO$", "Plaquemines SO", regex=True)
+                .str.replace(r"\bNo\b", "New Orleans PD", regex=True)
+                .str.replace(r"^St\. Martin$", "St. Martin SO")
+                .str.replace(r"^Tangipahoa$", "Tangipahoa SO", regex=True)
+                .str.replace(r"^Lake Charles$", "Lake Charles PD")
+                .str.replace(r"^St\. Tammany$", "St. Tammany SO", regex=True)
+                .str.replace(r"^Hammond$", "Hammond PD", regex=True)
+                .str.replace(r"^New Orleans$", "New Orleans PD")
+                .str.replace(r"^Lafayette City$", "Lafayette City PD", regex=True)
+                .str.replace(r"^Grand Isle$", "Grand Isle PD", regex=True)
+                .str.replace(r"^New Orleans Levee$", "New Orleans Levee PD", regex=True)
+                .str.replace(r"^Harbor PD$", "New Orleans Harbor PD", regex=True)
+                .str.replace(r"^Lafayette$", "Lafayette PD", regex=True)
+                .str.replace(r"^Louisiana State$", "Louisiana State PD", regex=True)
+                .str.replace(r"^Caddo Parish$", "Caddo SO", regex=True)
+            )
+    return df
+
+
+def drop_duplicates(df):
+    return df.drop_duplicates(subset=["first_name", "last_name"], keep="last")
+
+
+def generate_history_id(df):
+    stacked_agency_sr = df[
+        [
+            "agency",
+            "agency_1",
+            "agency_2",
+            "agency_3",
+            "agency_4",
+            "agency_5",
+            "agency_6",
+            "agency_7",
+            "agency_8",
+        ]
+    ].stack()
+
+    stacked_agency_df = stacked_agency_sr.reset_index().iloc[:, [0, 2]]
+    stacked_agency_df.columns = ["history_id", "agency"]
+
+    names_df = df[["first_name", "last_name", "middle_name"]].reset_index()
+    names_df = names_df.rename(columns={"index": "history_id"})
+
+    stacked_agency_df = stacked_agency_df.merge(names_df, on="history_id", how="right")
+
+    return stacked_agency_df
+
+
+def check_for_duplicate_uids(df):
+    uids = df.groupby(["uid"])["history_id"].agg(list).reset_index()
+
+    for row in uids["history_id"]:
+        unique = all(element == row[0] for element in row)
+        if unique:
+            continue
+        else:
+            raise ValueError("uid found in multiple history ids")
+
+    return df[~((df.agency.fillna("") == ""))]
+
+
+def generate_false_neg_id(df):
+    false_negs = [
+        ids == False for ids in df.duplicated(subset=["history_id"], keep=False)
+    ]
+    df.loc[:, "false_negatives"] = false_negs
+    df.loc[:, "false_negatives"] = df.astype(str).false_negatives.str.replace(
+        "False", "", regex=False
+    )
+
+    false_df = df.pipe(
+        gen_uid, ["false_negatives", "history_id"], "false_neg_id"
+    ).drop_duplicates(subset=["false_neg_id"], keep=False)
+    false_df = false_df[["false_neg_id", "history_id"]]
```

Comment:
The column should be named "switched_job" or something similar. I think this name is more descriptive and open this column up to more purposes. And you don't need to generate any id. Just keep it a boolean series:
```python
df.loc[:, 'switched_job'] = df.duplicated(subset=["history_id"], keep=False)
```

In the `cross_agency` script you are using `DissimilarFilter("false_neg_id")`, which attempts to filter out pairs that have the same `history_id` (`false_negatives` was also generated from `history_id` so in truth it all comes down to `history_id`). Doing this is not necessary because the records that we want to eliminate from the matching process all have only one `uid` associated with a `history_id`. Which mean it is only one row, not enough to form a pair. Not only that, I think it might wipe out true positives because rows with the same `history_id` (officers who switched job) will still have the same `false_neg_id`.

We don't need to filter at the pair level, we can simply filter at the record level. Add this line before cross_agency.py line 131 would be enough:
```python
per = discard_rows(per, per.switched_job, 'officers who has not switched job')
```

Reply:
Makes sense. Thanks.